modules/aws/master/master.tf

```diff
@@ -18,69 +15,85 @@ resource "aws_instance" "master" {
 
   associate_public_ip_address = false
   source_dest_check           = false
-  subnet_id                   = "${var.master_subnet_ids[count.index]}"
-  vpc_security_group_ids      = ["${aws_security_group.master.id}"]
+
+  subnet_id              = "${var.master_subnet_ids[count.index]}"
+  vpc_security_group_ids = ["${aws_security_group.master.id}"]
 
   root_block_device = {
     volume_type = "${var.volume_type}"
     volume_size = "${var.volume_size_root}"
   }
 
-  user_data = "${data.ignition_config.s3.rendered}"
+  user_data = "${element(data.ignition_config.s3.*.rendered, count.index)}"
 
   tags = "${merge(
     local.common_tags,
     map(
       "Name", "${var.cluster_name}-master${count.index}"
     )
   )}"
+
+  # we ignore changes, to avoid rolling all masters at once
+  # update is done via tainting masters
```

Comment:
This could be an improvement for the future may be a discussion for the SIG Infra.

We could put Masters under a separate ASG. So 3 AZs = 3 ASGs. When a new Launch Configuration is detected (a change in the user data) automatically a new ASG is created in TF. Then with a separate script we:

- drain each master
- terminate the instance. Automatically a new instance will be created (since we run in ASG) with a new LaunchConfiguration.

The process is automated and less error prone. WDTY?

Reply:
Yes, we already thought about this but the main problem with ASG is that if your instance dies, all EBS are gone, and this is a problematic for etcd cluster as its very easy to break cluster when some of your nodes are dead.

 with simple ec2 if the master goes down we still have ebs and rerunning terraform will start a new instance with the same data and we are 100% sure we wont break etcd cluster.

AWS quite often schedule hardware for maintenance or have other events that can cause your instance to be deleted which would in our case means that etcd data on one machine is lost.

In general, it would be nice to have  ASG only, but having a very reliable etcd cluster under ASG is very complex. There is a nice blog post about it and it recommends 9 ASGs and 9 nodes in the cluster ( 3 per each AZ).

Reply:
I see that's true. Thank you for the clarification. But would be an amazing improvement (and maybe not hard to implement) something like at the bootstrap of the master node look for any suitable etcd volume (if exists) to mount. Otherwise create a new one. And with this would not matter anymore if we run an EC2 instance inside or outside an ASG.

Reply:
You put the idea into a new issue and put it into sig-infra, if you have and idea how to do this reliably than sure let's do it.

Reply:
Going to do that ;-) thank you for the follow up on this.